src/main/java/seedu/toluist/dispatcher/CommandDispatcher.java

```diff
@@ -43,21 +44,41 @@
     private final EventsCenter eventsCenter = EventsCenter.getInstance();
     private final AliasTable aliasConfig = Config.getInstance().getAliasTable();
 
+    /**
+     * ArrayList to store previous commands entered since starting the application
+     */
+    private ArrayList<String> commandHistory;
+    protected static int historyPointer = 0;
+
     public CommandDispatcher() {
         super();
         aliasConfig.setReservedKeywords(getControllerKeywords());
+        commandHistory = new ArrayList<String>();
     }
 
     public void dispatch(Ui renderer, String command) {
+        recordCommand(command);
         String deAliasedCommand = aliasConfig.dealias(command);
         logger.info("De-aliased command to be dispatched: " + deAliasedCommand + " original command " + command);
 
         Controller controller = getBestFitController(renderer, deAliasedCommand);
         logger.info("Controller class to be executed: " + controller.getClass());
-        CommandResult feedbackToUser = controller.execute(deAliasedCommand);
+
+        CommandResult feedbackToUser;
+        if (controller.getClass().isInstance(new HistoryController(renderer))) {
+            feedbackToUser = ((HistoryController) controller).execute(deAliasedCommand, commandHistory);
+        } else {
+            feedbackToUser = controller.execute(deAliasedCommand);
+        }
+
         eventsCenter.post(new NewResultAvailableEvent(feedbackToUser.getFeedbackToUser()));
     }
 
+    private void recordCommand(String command) {
+        commandHistory.add(command);
+        historyPointer = commandHistory.size();
```

Comment:
If this line is the only reason why u need historyPointer, I think u can just remove this and keep calling commandHistory.size() whenever u need to get this variable?